File: navigation_filter_library/src/state.cpp
#include "state.hpp"

State::State(Eigen::Matrix<float, 12, 12> covariance) : covariance_(covariance)
{
    position_ << 0, 0, 0;
    velocity_ << 0, 0, 0;
    orientation_ << 0, 0, 0;
    angular_velocity_ << 0,0,0;
}

State::State()
{
    position_ << 0, 0, 0;
    velocity_ << 0, 0, 0;
    orientation_ << 0, 0, 0;
    angular_velocity_ << 0, 0, 0;

    covariance_.setIdentity();
    covariance_ *= 1e-8; // Initialize covariance to a small value
}

Eigen::Matrix<float, 12, 12> State::get_covariance()
{
    return covariance_;
}
Eigen::Vector<float, 12> State::get_state_vector()
{
    Eigen::Vector<float, 12> vector;
    vector << position_, velocity_, orientation_, angular_velocity_;
    return vector;
}
State::~State()
{
    // Destructor implementation if needed
}

File: sensor_fusion_modelling/include/state.py
import numpy as np


class State:

    def __init__(self, covariance: np.ndarray = None):
        self.position_ = np.zeros(3, dtype=np.float32)
        self.velocity_ = np.zeros(3, dtype=np.float32)
        self.orientation_ = np.zeros(3, dtype=np.float32)
        self.angular_velocity_ = np.zeros(3, dtype=np.float32)

        if covariance is not None:
            # equivalente a State(Eigen::Matrix<float,12,12> covariance)
            self.covariance_ = np.asarray(covariance, dtype=np.float32)
        else:
            # equivalente a State() default: identidade * 1e-8
            self.covariance_ = np.eye(12, dtype=np.float32) * 1e-8

    def get_covariance(self) -> np.ndarray:
        return self.covariance_

    def get_state_vector(self) -> np.ndarray:
        return np.concatenate([
            self.position_, self.velocity_,
            self.orientation_, self.angular_velocity_,
        ]).astype(np.float32)